### DISCUSSION TITLE
Having trouble with nvs subsystem on same54 processor. (Using the same54_xpro eval board)

### DISCUSSION THREAD

**Discussion by CarloATX** [zephyrproject-rtos/zephyr#82658] [Q&A] [answered]
Hi everyone, I'm running into issues with storage on the same54_xpro eval board. In my personal program I require to have some data be stored on the board through power cycles and what I find is that after I updated from zephyr v3.5 to v3.7, nvs storage is failing to persist between power cycles.

When I run this program `zephyr/samples/subsys/nvs/` on the eval board I get multiple warnings indicating issues with the `zephyr/drivers/flash/flash_sam0.c` driver:

![image](https://github.com/user-attachments/assets/739a0d2a-fdc0-498c-8bc8-e68f0a5a9ec2)

As you can see, the program does not find the data that was stored in the initial run continuing to "store" the info and rebooting the eval board expecting to see it on the next run.

I noticed there was a commit in v3.6 that added some changes [This pr](https://github.com/zephyrproject-rtos/zephyr/commit/debc65fa631d5497a85b0a418b96e1df323de0f3) . 

This is the only pr in recent history between v3.5 and v3.7 that I can point to that caused this behavior but I have not fully confirmed this on my end. I just know that if I run the sample nvs program, it fails on the eval board but if I run the test suite for flash/common, it passes.

I hope someone can provide guidance/clarity on this matter. Thank you for any assistance.

**Comment by Laczen** [comment]
Hi @CarloATX, the log shows warnings that it is trying to access flash outside of the available flash. This can only be due to a length error, it would be good to slightly modify the warn message in flash_sam0.c at line 106 to also report the length that was requested.

**Reply by CarloATX** [reply]
Thank you for the suggestion @Laczen .

The len is printed out to be `8`

![image](https://github.com/user-attachments/assets/ee07abaf-6a20-453c-8907-2336b1222a9f)

**Reply by Laczen** [reply]
This shouldn't be past the end of flash, could you verify also CONFiG_FLASH_SIZE * 1024? What is the flash size of your device?

**Reply by CarloATX** [reply]
In my build directory, config flash size is set as `1024`, which seems correct because the processor is same54p20a which has `1024 KB`.  

I also ran gdb yesterday on the program to print out the values for offset and length as well:
![gdb_samples_nvs](https://github.com/user-attachments/assets/8f9dd9cd-978b-4f47-bcee-262ebf832e17)
 I believe the offset is printed in decimal in the cmd window

**Reply by CarloATX** [reply]
I will run gdb once more to make sure that is the correct image. I may have run this for my personal program

**Reply by CarloATX** [reply]
> I will run gdb once more to make sure that is the correct image. I may have run this for my personal program

the picture is correct. this is what I get when I run the nvs sample program on the eval board

**Reply by Laczen** [reply]
The writing is at a too large value in your first report, it should be below 0x100000, how have you defined the offset in NVS?

**Reply by CarloATX** [reply]
I haven't written anything to define the offset for nvs. Currently I am running what is upstream on zephyr so the same54_xpro dts was unmodified from what is on zephyr 

for how the offset is defined/calculated, the flash_sam0.c driver loops over each lock region and adds the size of the lock region to the base flash address to set the address for the nvmctrl register. 
![image](https://github.com/user-attachments/assets/db62236f-c62d-4926-94ca-0159b5ec8fc3)

The next image is the set of macros that define the lock region sizes:
![image](https://github.com/user-attachments/assets/c5451f30-0b83-4ed2-8df0-a81fc6749764)

**Reply by Laczen** [reply]
Could you check the value of NVS_PARTITION_OFFSET, the range validity is checked at the beginning of a read.

**Reply by CarloATX** [reply]
![image](https://github.com/user-attachments/assets/53be3b8a-4def-43bd-b519-61b3d64868af)

it says 1032192 and when converted to hex, `0xFC000`.
according to the device tree, that is the start of the storage partition. I will step through more and see where there is any major changes

**Reply by Laczen** [reply]
Could you add some code to print out the NVS properties (especially sector_size). According to the datasheet this should be 8kB (if I read the datasheet correctly) but the driver might be returning 16kB instead.

**Reply by CarloATX** [reply]
I left the board at work for the weekend. I will be able to continue once I'm back. 
So far I really appreciate your attentiveness on this.

**Reply by CarloATX** [reply]
@Laczen Hello again. I am back and was able to add more prints for nvs_fs properties. 
![image](https://github.com/user-attachments/assets/b71f450e-9561-4b6e-8008-6d9646e7beeb)

**Reply by Laczen** [reply]
Hi @CarloATX, with the setup you can only use 2 sectors (the third sector is located outside of your flash), you either need to reduce the offset of the storage (by 8192 byte, so at 0xFA000) area or to modify the sample so that it only uses 2 sectors.

**Reply by CarloATX** [reply]
Ah that makes sense for running the sample nvs program at least. The errors for the addresses ending past flash are gone now. I decided to just modify the dts instead of modifying the program more.

Now I run back into the error where the flash driver is giving verify errors. The nvs sample loops for three runs and then crashes on the third. I had similar issues before which is what led me to think the flash driver is in a broken state

![image](https://github.com/user-attachments/assets/69a4993c-d452-4bad-aa28-8edc99fb8105)

Below are the full logs for the run:
[Putty Logs.txt](https://github.com/user-attachments/files/18064451/Putty.Logs.txt)

**Reply by Laczen** [reply]
Can you erase the area before starting the test (maybe you do this already)?

**Reply by CarloATX** [reply]
Yes, I made sure to delete the memory before running again.

**Reply by Laczen** [reply]
I see that you have enabled 1 byte write mode. Is it the same for the standard 8 byte write mode?

**Reply by CarloATX** [reply]
I actually haven't explicitly enabled the 1 byte write mode. Is there a specific config I can use to change the write mode to 8 byte or should I be modifying the program directly somehow?

**Reply by Laczen** [reply, marked as answer]
That is strange, in a previous log it was showing a write block size of 8, now 1. According  to the flash driver Kconfig it should default to 1 for certain devices. It should not change between compiles.
You can disable the option by setting CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES=n.

**Reply by CarloATX** [reply]
Hmm. There might be one thing that comes to mind that could cause this. There is this config `CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES=y` that is required for the sam0 soc and this might be modifying the write length. I may have added this in between compiles. as I was testing different combinations for building. I have to rebuild and confirm that this is the case however, it will take me a moment. right now I am building the nvs sample program on zephyr v4.0.0 but with the `flash_sam0.c` pr, reverted

**Reply by CarloATX** [reply]
> That is strange, in a previous log it was showing a write block size of 8, now 1. According to the flash driver Kconfig it should default to 1 for certain devices. It should not change between compiles. You can disable the option by setting CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES=n.

It looks like the issue was most likely an error on my part although I was sure I was having this issue before even setting the emulate byte pages flag. I am now not having issues with the nvs subsystem and my personal code is working as intended. 

setting `CONFIG_SOC_FLASH_SAM0_EMULATE_BYTE_PAGES=n` as you suggested resolved the issue. I am not sure why I did not notice this issue. I was so sure... 🤦‍♂️

**Reply by CarloATX** [reply]
Thank you so much for your help with this matter @Laczen
This is greatly appreciated that it was easily resolved with one config

**Reply by CarloATX** [reply]
on a side note, what is the emulate config used for now? Do you happen to know if there is any conflict with building mcuboot with this config and running an app that is built without this config?

Thanks again

**Reply by Laczen** [reply]
I don't really know what subsystems require the use of emulated byte write. There should not be a conflict between mcuboot and the app. However if there is a problem in the emulated byte write and mcuboot would use these mode (after recompilation) mcuboot might fail. Given your experience with the emulated byte write I would suspect there is a problem with this mode and it is not detected by the flash test.
